src/main/java/com/aws/greengrass/util/platforms/windows/WindowsRunasProcess.java

```diff
@@ -0,0 +1,450 @@
+     * @param username name of the account
+     */
+    public WindowsRunasProcess(String domain, String username) {
+        super();
+        this.domain = domain;
+        this.username = username;
+    }
+
+    /**
+     * Set env vars in addition to the user's own existing env vars.
+     *
+     * @param envs a map of env vars to set
+     */
+    public void setEnv(Map<String, String> envs) {
+        additionalEnv.putAll(envs);
+    }
+
+    public synchronized void setCurrentDirectory(String dir) {
+        currentDirectory = dir;
+    }
+
+    /**
+     * Start the process as given user.
+     *
+     * @param command command line to run
+     * @throws IOException if failed to run subprocess
+     */
+    public synchronized void start(String command) throws IOException {
+        WinNT.HANDLEByReference processTokenHandle = new WinNT.HANDLEByReference();
+        if (!Advapi32.INSTANCE.OpenProcessToken(Kernel32.INSTANCE.GetCurrentProcess(), Kernel32.TOKEN_ALL_ACCESS,
+                processTokenHandle)) {
+            throw lastErrorProcessCreationException("OpenProcessToken");
+        }
+
+        if (isService(processTokenHandle)) {
+            processAsUser(processTokenHandle, command);
+        } else {
+            processWithLogon(command);
+        }
+
+        Kernel32Util.closeHandleRefs(processTokenHandle);
+    }
+
+    private void processWithLogon(final String command) throws IOException {
+        char[] password = getPassword();
+        WinBase.STARTUPINFO startupInfo = initPipesAndStartupInfo();
+        // TODO set additional env vars
+        procInfo = new WinBase.PROCESS_INFORMATION();
+
+        if (!ProcAdvapi32.INSTANCE.CreateProcessWithLogonW(username, domain, password, Advapi32.LOGON_WITH_PROFILE,
+                null, command, PROCESS_CREATION_FLAGS, computeEnvironmentBlock(), currentDirectory, startupInfo,
+                procInfo)) {
+            throw lastErrorProcessCreationException("CreateProcessWithLogonW");
+        }
+        Arrays.fill(password, NULL_CHAR);
+
+        pid = Kernel32.INSTANCE.GetProcessId(procInfo.hProcess);
+        if (pid == 0) {
+            throw lastErrorProcessCreationException("GetProcessId");
+        }
+
+        redirectStreams();
+    }
+
+    private void processAsUser(final WinNT.HANDLEByReference processTokenHandle, final String command)
+            throws IOException {
+        enablePrivileges(processTokenHandle.getValue(), Kernel32.SE_TCB_NAME, Kernel32.SE_ASSIGNPRIMARYTOKEN_NAME);
+
+        char[] password = getPassword();
+        // LogonUser
+        boolean logonSuccess = false;
+        WinNT.HANDLEByReference userTokenHandle = new WinNT.HANDLEByReference();
+        int[] logonTypes =
+                {Kernel32.LOGON32_LOGON_INTERACTIVE, Kernel32.LOGON32_LOGON_SERVICE, Kernel32.LOGON32_LOGON_BATCH};
+        for (int logonType : logonTypes) {
+            if (ProcAdvapi32.INSTANCE.LogonUser(username, domain, password, logonType,
```

Comment:
Is this safe to run in parallel? Think about this for all the win api calls here too. We may create multiple processes all at once and we don't want them to conflict

Reply:
Added sync to Advapi32.INSTANCE for the entire flow.

Reply:
Well if it isn't necessary, don't add it.

Reply:
I think we should. At least the last error can be messed up if APIs are called in parallel.